### ISSUE TITLE
Add some kind of personal signature to de-odexed apps

### ISSUE THREAD

**Issue by mfonville** [opengapps/opengapps#92]
When an apk is imported, it should be checked whether it is valid, not malicious or broken. E.g. like happened with last x86 tts.
The import script should verify the meta-inf information using the jar signature verification mechanism. And compare the certificate against Google's.
If an apk does not pass, if should not be imported.
We might also want to run such a test on the buildbot server.

**Comment by mfonville** [comment]
An idea is btw, to also enable such checks during the build process.

**Comment by Lekensteyn** [comment]
Are you aware that `jarsigner -verify` can be used for this?

First create a keystore from a known trusted key. `-text` is optional, but useful to have a human-readable block of text in addition to the PEM-encoded data. The `-alias NAME` must be unique for every imported certificate:

```sh
bsdtar xOf CalendarProvider.apk META-INF/CERT.RSA |
openssl pkcs7 -inform DER -print_certs -text > cert.pem
keytool -importcert -file cert.pem -keystore keystore -storepass changeit -noprompt -alias cert1

# alternative invocation without creating cert.pem file
bsdtar xOf GoogleServicesFramework.apk META-INF/CERT.RSA |
openssl pkcs7 -inform DER -print_certs |
keytool -importcert -keystore keystore -storepass changeit -noprompt -alias cert2
```

Then you can verify other APKs with:

```sh
jarsigner -verify -keystore keystore -verbose:summary CalendarProvider.apk
```

Example **good** output (note that classes.dex is not signed, see also https://github.com/Lekensteyn/make-gapps-zip/issues/3):
```
smk     8528 Tue Sep 03 11:53:26 CEST 2013 AndroidManifest.xml (and 4 more)
s k      453 Tue Sep 03 11:53:26 CEST 2013 META-INF/MANIFEST.MF
         498 Tue Sep 03 11:53:26 CEST 2013 META-INF/CERT.SF (and 1 more)
       92064 Tue Sep 03 11:53:26 CEST 2013 classes.dex

  s = signature was verified 
  m = entry is listed in manifest
  k = at least one certificate was found in keystore
  i = at least one certificate was found in identity scope

jar verified.

Warning: 
This jar contains unsigned entries which have not been integrity-checked. 
This jar contains signatures that does not include a timestamp. Without a timestamp, users may not be able to validate this jar after the signer certificate's expiration date (2038-01-19) or after any future revocation date.

Re-run with the -verbose and -certs options for more details.
```

Example **bad** output (only differences are shown for brevity, note that classes.dex was signed this time):
<pre>
<b>s</b>        846 Mon Aug 17 23:30:16 CEST 2015 META-INF/MANIFEST.MF
         793 Mon Aug 17 23:32:32 CEST 2015 META-INF/CERT.SF (and 1 more)
<b>sm</b>      8528 Tue Sep 03 11:53:26 CEST 2013 AndroidManifest.xml (and 5 more)
...
Warning: 
<b>This jar contains entries whose certificate chain is not validated.</b>
...
</pre>

The contents of cert.pem (for reference; first certificate is the newer (2013) certificate from CalendarProvider, second one is an older (2008) certificate from GoogleServicesFramework):
```
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number: 1378230807839 (0x140e4f52d1f)
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=US, ST=California, L=Mountain View, O=Google Inc., OU=Android, CN=Android
        Validity
            Not Before: Sep  3 17:53:27 2013 GMT
            Not After : Jan 19 03:14:07 2038 GMT
        Subject: C=US, ST=California, L=Mountain View, O=Google Inc., OU=Android, CN=Android
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:c8:3c:64:6f:c4:b6:4b:83:eb:81:14:4f:e8:85:
                    a8:dc:ba:b4:33:04:87:75:42:a4:2d:98:7d:ef:ec:
                    2b:9c:6a:2b:7a:81:e1:60:c6:1c:d3:7a:cc:46:a9:
                    3c:40:81:c2:94:d7:5c:8a:2f:98:26:4d:21:00:b0:
                    ca:78:1e:b5:17:33:af:fb:98:53:b0:7d:91:77:09:
                    63:23:57:c1:b2:f7:fc:20:35:99:10:79:98:2a:32:
                    97:1b:0d:1c:8d:13:ef:62:30:a4:8a:31:14:d8:2f:
                    2e:87:b9:81:2a:af:0b:ab:11:c3:f8:eb:fa:cb:e8:
                    af:f0:61:63:14:da:93:f8:c1:25:1f:40:f6:1e:c3:
                    20:0c:00:19:9a:4b:02:95:ad:a7:8c:80:e1:e3:0f:
                    c2:0b:68:f0:64:9a:0d:0e:6b:e3:2b:ee:e2:54:0e:
                    09:f8:af:69:17:04:14:b7:37:f2:12:67:00:c7:ea:
                    9b:06:52:49:b1:04:c3:e0:98:b2:5d:ec:15:1e:0e:
                    d3:f6:34:8a:9a:21:2e:b3:7f:11:37:d5:9e:f5:25:
                    4c:34:f6:6f:99:c6:57:ed:d3:7b:44:18:27:82:89:
                    92:c2:8d:27:76:8e:48:a1:46:82:88:00:bd:32:04:
                    4f:c5:84:33:a3:5c:ff:55:b4:67:aa:7a:1b:96:96:
                    4a:11
                Exponent: 65537 (0x10001)
    Signature Algorithm: sha1WithRSAEncryption
         0c:02:2e:ab:f5:55:7e:a6:8d:cf:a8:c1:fe:49:fd:4e:e1:24:
         c4:09:b9:ce:f8:92:e0:6a:48:8b:40:88:56:96:23:4f:fd:25:
         72:90:de:0b:3e:d3:b9:9b:5e:45:73:dc:e6:e6:8b:85:5c:f7:
         dd:49:81:95:bb:ee:d5:ca:5f:b9:70:26:94:98:a4:b5:da:03:
         2a:75:98:26:8b:9a:42:f6:7f:2c:75:e6:68:0e:a1:4d:d8:47:
         f9:ca:02:a9:db:2a:26:68:6f:3e:62:c0:ee:4a:af:83:14:55:
         e3:9f:21:6c:eb:bb:59:0a:fe:2a:92:18:21:2f:c7:f5:79:ec:
         63:1e:d6:1c:b3:a9:b5:15:f5:58:6e:d5:65:9f:19:16:26:04:
         c1:7c:4b:76:fc:95:2c:6c:05:a3:a9:aa:0f:74:de:54:48:75:
         d1:64:bd:7d:64:58:f0:43:e7:19:67:7f:b1:fe:92:13:bd:4b:
         f3:39:6f:c0:21:e3:0e:dc:c1:aa:be:cd:73:5e:8b:c2:3b:2a:
         72:f5:82:59:10:fe:39:a9:e6:a1:c9:a5:e8:72:a9:6a:97:c4:
         fb:02:e8:66:80:77:d9:06:76:45:34:77:97:14:92:3c:90:e4:
         e7:21:8b:d4:11:44:f4:b0:c6:5a:91:a8:43:00:a8:cd:11:96:
         4f:4f:80:0e
-----BEGIN CERTIFICATE-----
MIIDXTCCAkcCBgFA5PUtHzALBgkqhkiG9w0BAQUwdDELMAkGA1UEBhMCVVMxEzAR
BgNVBAgTCkNhbGlmb3JuaWExFjAUBgNVBAcTDU1vdW50YWluIFZpZXcxFDASBgNV
BAoTC0dvb2dsZSBJbmMuMRAwDgYDVQQLEwdBbmRyb2lkMRAwDgYDVQQDEwdBbmRy
b2lkMB4XDTEzMDkwMzE3NTMyN1oXDTM4MDExOTAzMTQwN1owdDELMAkGA1UEBhMC
VVMxEzARBgNVBAgTCkNhbGlmb3JuaWExFjAUBgNVBAcTDU1vdW50YWluIFZpZXcx
FDASBgNVBAoTC0dvb2dsZSBJbmMuMRAwDgYDVQQLEwdBbmRyb2lkMRAwDgYDVQQD
EwdBbmRyb2lkMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyDxkb8S2
S4PrgRRP6IWo3Lq0MwSHdUKkLZh97+wrnGoreoHhYMYc03rMRqk8QIHClNdcii+Y
Jk0hALDKeB61FzOv+5hTsH2RdwljI1fBsvf8IDWZEHmYKjKXGw0cjRPvYjCkijEU
2C8uh7mBKq8LqxHD+Ov6y+iv8GFjFNqT+MElH0D2HsMgDAAZmksCla2njIDh4w/C
C2jwZJoNDmvjK+7iVA4J+K9pFwQUtzfyEmcAx+qbBlJJsQTD4JiyXewVHg7T9jSK
miEus38RN9We9SVMNPZvmcZX7dN7RBgngomSwo0ndo5IoUaCiAC9MgRPxYQzo1z/
VbRnqnoblpZKEQIDAQABMAsGCSqGSIb3DQEBBQOCAQEADAIuq/VVfqaNz6jB/kn9
TuEkxAm5zviS4GpIi0CIVpYjT/0lcpDeCz7TuZteRXPc5uaLhVz33UmBlbvu1cpf
uXAmlJiktdoDKnWYJouaQvZ/LHXmaA6hTdhH+coCqdsqJmhvPmLA7kqvgxRV458h
bOu7WQr+KpIYIS/H9XnsYx7WHLOptRX1WG7VZZ8ZFiYEwXxLdvyVLGwFo6mqD3Te
VEh10WS9fWRY8EPnGWd/sf6SE71L8zlvwCHjDtzBqr7Nc16LwjsqcvWCWRD+Oanm
ocml6HKpapfE+wLoZoB32QZ2RTR3lxSSPJDk5yGL1BFE9LDGWpGoQwCozRGWT0+A
Dg==
-----END CERTIFICATE-----
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 14042372374541250701 (0xc2e08746644a308d)
    Signature Algorithm: md5WithRSAEncryption
        Issuer: C=US, ST=California, L=Mountain View, O=Google Inc., OU=Android, CN=Android
        Validity
            Not Before: Aug 21 23:13:34 2008 GMT
            Not After : Jan  7 23:13:34 2036 GMT
        Subject: C=US, ST=California, L=Mountain View, O=Google Inc., OU=Android, CN=Android
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:ab:56:2e:00:d8:3b:a2:08:ae:0a:96:6f:12:4e:
                    29:da:11:f2:ab:56:d0:8f:58:e2:cc:a9:13:03:e9:
                    b7:54:d3:72:f6:40:a7:1b:1d:cb:13:09:67:62:4e:
                    46:56:a7:77:6a:92:19:3d:b2:e5:bf:b7:24:a9:1e:
                    77:18:8b:0e:6a:47:a4:3b:33:d9:60:9b:77:18:31:
                    45:cc:df:7b:2e:58:66:74:c9:e1:56:5b:1f:4c:6a:
                    59:55:bf:f2:51:a6:3d:ab:f9:c5:5c:27:22:22:52:
                    e8:75:e4:f8:15:4a:64:5f:89:71:68:c0:b1:bf:c6:
                    12:ea:bf:78:57:69:bb:34:aa:79:84:dc:7e:2e:a2:
                    76:4c:ae:83:07:d8:c1:71:54:d7:ee:5f:64:a5:1a:
                    44:a6:02:c2:49:05:41:57:dc:02:cd:5f:5c:0e:55:
                    fb:ef:85:19:fb:e3:27:f0:b1:51:16:92:c5:a0:6f:
                    19:d1:83:85:f5:c4:db:c2:d6:b9:3f:68:cc:29:79:
                    c7:0e:18:ab:93:86:6b:3b:d5:db:89:99:55:2a:0e:
                    3b:4c:99:df:58:fb:91:8b:ed:c1:82:ba:35:e0:03:
                    c1:b4:b1:0d:d2:44:a8:ee:24:ff:fd:33:38:72:ab:
                    52:21:98:5e:da:b0:fc:0d:0b:14:5b:6a:a1:92:85:
                    8e:79
                Exponent: 3 (0x3)
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                C7:7D:8C:C2:21:17:56:25:9A:7F:D3:82:DF:6B:E3:98:E4:D7:86:A5
            X509v3 Authority Key Identifier: 
                keyid:C7:7D:8C:C2:21:17:56:25:9A:7F:D3:82:DF:6B:E3:98:E4:D7:86:A5
                DirName:/C=US/ST=California/L=Mountain View/O=Google Inc./OU=Android/CN=Android
                serial:C2:E0:87:46:64:4A:30:8D

            X509v3 Basic Constraints: 
                CA:TRUE
    Signature Algorithm: md5WithRSAEncryption
         6d:d2:52:ce:ef:85:30:2c:36:0a:aa:ce:93:9b:cf:f2:cc:a9:
         04:bb:5d:7a:16:61:f8:ae:46:b2:99:42:04:d0:ff:4a:68:c7:
         ed:1a:53:1e:c4:59:5a:62:3c:e6:07:63:b1:67:29:7a:7a:e3:
         57:12:c4:07:f2:08:f0:cb:10:94:29:12:4d:7b:10:62:19:c0:
         84:ca:3e:b3:f9:ad:5f:b8:71:ef:92:26:9a:8b:e2:8b:f1:6d:
         44:c8:d9:a0:8e:6c:b2:f0:05:bb:3f:e2:cb:96:44:7e:86:8e:
         73:10:76:ad:45:b3:3f:60:09:ea:19:c1:61:e6:26:41:aa:99:
         27:1d:fd:52:28:c5:c5:87:87:5d:db:7f:45:27:58:d6:61:f6:
         cc:0c:cc:b7:35:2e:42:4c:c4:36:5c:52:35:32:f7:32:51:37:
         59:3c:4a:e3:41:f4:db:41:ed:da:0d:0b:10:71:a7:c4:40:f0:
         fe:9e:a0:1c:b6:27:ca:67:43:69:d0:84:bd:2f:d9:11:ff:06:
         cd:bf:2c:fa:10:dc:0f:89:3a:e3:57:62:91:90:48:c7:ef:c6:
         4c:71:44:17:83:42:f7:05:81:c9:de:57:3a:f5:5b:39:0d:d7:
         fd:b9:41:86:31:89:5d:5f:75:9f:30:11:26:87:ff:62:14:10:
         c0:69:30:8a
-----BEGIN CERTIFICATE-----
MIIEQzCCAyugAwIBAgIJAMLgh0ZkSjCNMA0GCSqGSIb3DQEBBAUAMHQxCzAJBgNV
BAYTAlVTMRMwEQYDVQQIEwpDYWxpZm9ybmlhMRYwFAYDVQQHEw1Nb3VudGFpbiBW
aWV3MRQwEgYDVQQKEwtHb29nbGUgSW5jLjEQMA4GA1UECxMHQW5kcm9pZDEQMA4G
A1UEAxMHQW5kcm9pZDAeFw0wODA4MjEyMzEzMzRaFw0zNjAxMDcyMzEzMzRaMHQx
CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpDYWxpZm9ybmlhMRYwFAYDVQQHEw1Nb3Vu
dGFpbiBWaWV3MRQwEgYDVQQKEwtHb29nbGUgSW5jLjEQMA4GA1UECxMHQW5kcm9p
ZDEQMA4GA1UEAxMHQW5kcm9pZDCCASAwDQYJKoZIhvcNAQEBBQADggENADCCAQgC
ggEBAKtWLgDYO6IIrgqWbxJOKdoR8qtW0I9Y4sypEwPpt1TTcvZApxsdyxMJZ2JO
Rland2qSGT2y5b+3JKkedxiLDmpHpDsz2WCbdxgxRczfey5YZnTJ4VZbH0xqWVW/
8lGmPav5xVwnIiJS6HXk+BVKZF+JcWjAsb/GEuq/eFdpuzSqeYTcfi6idkyugwfY
wXFU1+5fZKUaRKYCwkkFQVfcAs1fXA5V+++FGfvjJ/CxURaSxaBvGdGDhfXE28LW
uT9ozCl5xw4Yq5OGazvV24mZVSoOO0yZ31j7kYvtwYK6NeADwbSxDdJEqO4k//0z
OHKrUiGYXtqw/A0LFFtqoZKFjnkCAQOjgdkwgdYwHQYDVR0OBBYEFMd9jMIhF1Yl
mn/Tgt9r45jk14alMIGmBgNVHSMEgZ4wgZuAFMd9jMIhF1Ylmn/Tgt9r45jk14al
oXikdjB0MQswCQYDVQQGEwJVUzETMBEGA1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UE
BxMNTW91bnRhaW4gVmlldzEUMBIGA1UEChMLR29vZ2xlIEluYy4xEDAOBgNVBAsT
B0FuZHJvaWQxEDAOBgNVBAMTB0FuZHJvaWSCCQDC4IdGZEowjTAMBgNVHRMEBTAD
AQH/MA0GCSqGSIb3DQEBBAUAA4IBAQBt0lLO74UwLDYKqs6Tm8/yzKkEu116FmH4
rkaymUIE0P9KaMftGlMexFlaYjzmB2OxZyl6euNXEsQH8gjwyxCUKRJNexBiGcCE
yj6z+a1fuHHvkiaai+KL8W1EyNmgjmyy8AW7P+LLlkR+ho5zEHatRbM/YAnqGcFh
5iZBqpknHf1SKMXFh4dd239FJ1jWYfbMDMy3NS5CTMQ2XFI1MvcyUTdZPErjQfTb
Qe3aDQsQcafEQPD+nqActifKZ0Np0IS9L9kR/wbNvyz6ENwPiTrjV2KRkEjH78ZM
cUQXg0L3BYHJ3lc69Vs5Ddf9uUGGMYldX3WfMBEmh/9iFBDAaTCK
-----END CERTIFICATE-----
```
(from the apks shipped with the Nexus 5, version 5.1.1 (LMY48I))

**Comment by mfonville** [comment]
Yes, I know about jarsigner (could just not remember the command here without a pc to check). But we might also have to add some checks that jarsigner does not.
E.g. I believe (but cannot check right now) that jarsigner does not look for/checks files that are NOT in the jar, but are mentioned in the META-INF

**Comment by Lekensteyn** [comment]
Confirmed, jarsigner only checks existing files. For that second case, you could write additional logic to extract the manifest file and read each `Name:` line.

Note that lines can be folded (like mail headers):
```
Name: res/layout/setup_default_list_item.xml
SHA1-Digest: 3f4155WqM1z/OfQWEVQ1JS+J9EQ=

Name: res/drawable-xhdpi-v4/notification_system_update_download_failur
 e.png
SHA1-Digest: wJ82YjmUe5+492Tk/2aMpFAVVKk=

Name: res/drawable-xxhdpi-v4/btn_dir_next_disabled_light.png
SHA1-Digest: KIzHD5xM1JwE8QTLVelI5iFbVyY=
```

**Comment by mfonville** [comment]
@Lekensteyn could you check my current implementation? I know it is not really the best code, and it all still depends on some assumptions how Google's certificates do look like e.g.
But do you think it should be enough to verify the integrity of APKs? e.g. that one could not add a malicious classes.dex and still be able to add as if it were from Google?

**Comment by Lekensteyn** [comment]
If `jarsigner` still exits with 1 while there are files with no signature in the keystore, then the current approach is broken. Can you verify the behavior against the example above?

**Comment by mfonville** [comment]
@Lekensteyn I assume you mean 'still exits with 0', right?
Because if an APK contains even a single file that is not present in the manifest and/or is not signed by a key from the keystore it with exits >=1.

**Comment by mfonville** [comment]
Btw, I don't know if an APK can have multiple certificates in the CERT.RSA, if this is possible, I don't know the current behaviour of the script...

**Comment by Lekensteyn** [comment]
Ah, the `-strict` option is the key here. With this option, items signed with a different key indeed produce a non-zero exit code. Additional items that are not signed also generate an error, so you do not have to check for them either. This can be tested by appending a file with `zip foo.apk dummy.txt` and then use `jarsigner` with the `-strict` flag.

There can only be one certificate in CERT.RSA AFAIK. It is however possible that more `FOO.RSA` and `FOO.SF` files exist in a jar (meaning that it got signed by multiple parties).

**Comment by mfonville** [comment]
ok, so about the CERT.RSA, the script would never import 'by accident' a non-Google key in the importcert function. And the verify method always demands that it is signed by a key from the keystore, ergo a Google key, and thus signing with any other key would still not match the keystore and be rejected by the script.

I wonder though, do jarsigner and keytool directly correctly verify the whole keychain using some root certificates? Otherwise one could generate a self-signed certificate and still sign the APK pretending to be Google, right? Do you have more knowledge about this @Lekensteyn ?

**Comment by mc1100** [comment]
My local scripts/certificates/opengapps.keystore changed after I used "add_sourceapp.sh" script. After I run the command "keytool -list -storepass opengapps -keystore opengapps.keystore" it proved one key got duplicated in the keystore:

Keystore type: JKS
Keystore provider: SUN

Your keystore contains 17 entries
...
Certificate fingerprint (SHA1): 38:91:8A:45:3D:07:19:93:54:F8:B1:9A:F0:5E:C6:56:2C:ED:57:88
serial, 7.9.2015, trustedCertEntry, 
Certificate fingerprint (SHA1): 38:91:8A:45:3D:07:19:93:54:F8:B1:9A:F0:5E:C6:56:2C:ED:57:88
15616810052367386195, 4.9.2015, trustedCertEntry, 
...

**Comment by mfonville** [comment]
Hmm, it is correct that it should change, because when importing new APKs, updated certificates are automatically added. But a completely identical key should not be re-added. I do see that two fields differ in your copy-paste, though?

**Comment by mc1100** [comment]
I am working on Mac OS X, so maybe some tools work different. E. g. I needed to replace original "find"  with homebrew otherwise make would not work. So if this is only me having this issue I will try to find the root cause.

**Comment by Lekensteyn** [comment]
@mfonville No Root CA in use and self-signed (CA) certificates can indeed "impersonate" Google. 

Here is an example where I used a generated certificate which passes your regex (subject `/C=US/ST=CA/L=Mountain View/O=Google Inc/OU=Android/CN=fake`):
```
+ jarsigner -verify -keystore keystore -verbose:summary -strict -certs test.apk

s k   348901 Mon Sep 07 23:14:32 CEST 2015 META-INF/MANIFEST.MF

      X.509, CN=fake, OU=Android, O=Google Inc, L=Mountain View, ST=CA, C=US (cert)
      [certificate will expire on 07/10/15 23:04]

      256234 Mon Sep 07 23:14:32 CEST 2015 META-INF/CERT.SF (and 1 more)

      (Signature related entries)

smk    61892 Thu Aug 21 17:13:34 CEST 2008 AndroidManifest.xml (and 2158 more)

      X.509, CN=fake, OU=Android, O=Google Inc, L=Mountain View, ST=CA, C=US (cert)
      [certificate will expire on 07/10/15 23:04]


  s = signature was verified 
  m = entry is listed in manifest
  k = at least one certificate was found in keystore
  i = at least one certificate was found in identity scope

jar verified.

Warning: 
This jar contains entries whose signer certificate will expire within six months. 
This jar contains signatures that does not include a timestamp. Without a timestamp, users may not be able to validate this jar after the signer certificate's expiration date (2015-10-07) or after any future revocation date.
+ echo 0
0
```

**Comment by mfonville** [comment]
I think I will then change the current procedure and do not automatic import any signatures within the script. But will manage them myself in the repo.

I wonder how the manual APK installation on an Android device handles this though. Since I would expect them to do some kind of APK signature checking. I do know that packages which claim e.g. permissions to a common resource, need to be signed with the same key (family?).
Without any CA the whole signature checking on Android is kind of bullshit, since anyone could make their own false signatures and drop their malicious code.

**Comment by mfonville** [comment]
Wrote something about this quest: https://plus.google.com/+OpengappsOrg/posts/LvpqiQxEhyg

**Comment by mc1100** [comment]
I checked the problem with the duplicate entry in the scripts/certificates/opengapps.keystore. There are different outputs from the openssl command. Sometimes 

$ unzip -p com.android.vending.apk META-INF/CERT.RSA | openssl pkcs7 -inform DER -print_certs -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            c2:e0:87:46:64:4a:30:8d
...

and sometimes

$unzip -p com.google.android.apps.docs.editors.docs.apk META-INF/CERT.RSA | openssl pkcs7 -inform DER -print_certs -text
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number: 1228183678 (0x4934987e)
...

Please, notice the new line right after the "Serial Number:" field in the former output. Unfortunately, only the later output format is expected in the function importcert() in the scripts/inc.sourceshelper.sh script. I do not know if this is openssl command line tool related or this is expected output for different certificate versions. Just for the reference, my openssl version:

$ openssl version
OpenSSL 0.9.8zg 14 July 2015

In any case I can post corrected alias retrieval.

**Comment by mfonville** [comment]
@mc1100 if you have a quick fix for the alias retrieval, that would indeed be welcome :-)

**Comment by mc1100** [comment]
I have just created pull request for this. You can check it.

**Comment by mfonville** [comment]
For now this is implemented as good as we can do. In general we only add packages that are 100% signed (not de-odexed) from Play Store and Google's Android One devices.
The keystore is relatively complete and is now manually managed from the main repository.
I will rename the bug to still find a solution to for the personall de-odexed apps.

**Comment by mfonville** [comment]
I forfeit the de-odexed signing plan. I considered the various options available like GPG signing of each APK, but none of them removes the single entry-point of possible security breakage: malicious access to the repo and people trusting me and my account.

Any malicious attacker could always, at the moment they would have access to my account, adjust the shell code, the keystore or any of those signatures provided.

Rather than providing a false sense of security, I would rather just plainly indicate that those APKs are considered to fail the verification test and should be treated as such. And to keep the the strongest incentive for all packagers to find fully signed APKs and to prefer those above manually de-odexing.
